clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -734,11 +740,11 @@ private ClusterResourceListeners configureClusterResourceListeners(Serializer<K>
     private int partition(ProducerRecord<K, V> record, byte[] serializedKey, byte[] serializedValue, Cluster cluster) {
         Integer partition = record.partition();
         if (partition != null) {
-            List<PartitionInfo> partitions = cluster.partitionsForTopic(record.topic());
-            int lastPartition = partitions.size() - 1;
+            int lastPartition = cluster.partitionsForTopic(record.topic()).size() - 1;
             // they have given us a partition, use it
             if (partition < 0 || partition > lastPartition) {
-                throw new IllegalArgumentException(String.format("Invalid partition given with record: %d is not in the range [0...%d].", partition, lastPartition));
+                throw new UnknownTopicOrPartitionException(
```

Comment:
I think I slightly prefer to keep the exception generic for now. My reasoning is that due to KIP-4, we are going to be changing the client's behavior in some ways when it encounters unknown topics or partitions. For example, instead of retrying indefinitely if a topic doesn't exist, the producer will either use the `CreateTopics` API directly or raise an exception. So I think I'd rather not expose a new exception type to the user which may tie our hands a little when it comes time to make those changes. Does that seem reasonable?